### ISSUE TITLE
Question about the caculation of cross-correlation

### ISSUE THREAD

**Issue by Achuan-2** [HelmchenLabSoftware/Centripetal_propagation_astrocytes#1]
![Clip_2024-04-18_23-15-50](https://github.com/HelmchenLabSoftware/Centripetal_propagation_astrocytes/assets/60436214/31c10b47-bd4b-4620-8c86-0f6bd8947263)
1. Why not use the built-in `xcorr` to calculate cross-correlation
2. I think  `delay = xi - max_delay`  should be `delay = xi - max_delay-1;` 
    Assuming the maximum delay is `kk=0`, then index corresponding to the cross_correlation is `max_delay+1`, but the calculated delay is -1
example code : I set `calculated_delay = xi - max_delay - 1;`
```matlab

signal_X = randn(100,1);  
t = 0:100-1;
delay_samples = 10;
signal_Y = circshift(signal_X, [delay_samples, 0]); % Signal Y is a delayed version of signal X

% Compute cross-correlation
max_delay = 20; 
cross_correlation = zeros(2*max_delay+1,1);
for kk = -max_delay:max_delay
    X = signal_X;
    Y = circshift(signal_Y,[kk, 0]);
    if kk >= 0
        cross_correlation(kk+max_delay+1) = corr(X(kk+1:end),Y(kk+1:end));
    else
        cross_correlation(kk+max_delay+1) = corr(X(1:end+kk),Y(1:end+kk));
    end
end

% Find the peak of cross-correlation
[ix,xi] = max(cross_correlation);
calculated_delay = xi - max_delay - 1; % Indexing starts from 1, so subtract 1
calculated_delay = -calculated_delay;
% Print actual and calculated delay
fprintf('Actual delay: %d \n', delay_samples);
fprintf('Calculated delay: %d \n', calculated_delay);

% Visualization
figure;
subplot(2,1,1);
plot(t, signal_X);
title('Signal X');
subplot(2,1,2);
plot(t, signal_Y);
title('Signal Y (Delayed)');

figure;
plot(-max_delay:max_delay, cross_correlation);
title('Cross-Correlation');
xlabel('Delay (Sample Points)');
ylabel('Correlation Coefficient');

```
>Actual delay: 10 
>Calculated delay: 10

**Comment by PTRRupprecht** [comment]
Hi @Achuan-2,

Thank you very much for bringing up these issues! You are definitely right about the indexing error when computing the delay. Thanks for pointing it out, and extra kudos for the instructive example! I've already fixed the problem. I guess that I introduced the problem when refactoring the code for Github.

About your second point, which is also a very valid and justified question: Yes, one could use the `xcorr` function and speed up the computation a lot. There are two reasons why I kept the loop together with the standard `corr` function. First, from my experience, correlation functions are kind of scary for many people who don't have a physics or engineering background. Therefore, using the well-know `corr` function together with a manually introduced temporal shift provides an easier access to understanding the code. Second, the `xcorr` function can be used with a lot of different normalization arguments. To stay consistent with the idea of just doing "temporally shifted Pearson's correlation", I was not fully sure which normalization would have to be applied. Would you have a recommendation on which arguments to use with the `xcorr` function in this context that would work well both with very long and very short recording durations?

Best wishes,
Peter

**Comment by Achuan-2** [comment]
To be honest, I am not familiar with `xcorr`, that's why I asked this question. Maybe I need to use it to test some data. The normalization argument of `xcorr` does seem important.

And I want to express my appreciation for your recent article, which I found to be both insightful and thought-provoking. The concept of "multi-timepoint, dilated linear regression" particularly caught my attention, and I believe it holds significant potential for application within the field of neuroscience. But,I don't quite understand the specific process of the regression. The following is the code I generated using GPT. Is this code correct? 
```matlab
function [beta, y_pred] = dilated_linear_regression(X, y, maxDilation)
    % X is the matrix of regressor variables (time by variables)
    % y is the vector of dependent variables (time by 1)
    % maxDilation is the maximum dilation factor to consider (e.g., 512 for ±17s at 30Hz)
    
    % Get the number of timepoints and regressors
    [numTimepoints, numRegressors] = size(X);
    
    % Initialize the new design matrix with dilated regressors
    X_dilated = zeros(numTimepoints, numRegressors * log2(maxDilation));
    
    % Fill the design matrix with dilated regressors
    colIdx = 1;
    for dilation = 1:log2(maxDilation)
        binSize = 2^(dilation - 1);
        for reg = 1:numRegressors
            for t = 1:numTimepoints
                binStart = max(1, t - binSize);
                binEnd = min(numTimepoints, t + binSize);
                X_dilated(t, colIdx) = mean(X(binStart:binEnd, reg));
            end
            colIdx = colIdx + 1;
        end
    end
    
    % Perform linear regression
    beta = regress(y, X_dilated);
    
    % Predict the dependent variable
    y_pred = X_dilated * beta;
end

``` 

I would be most grateful if you could share the associated code, perhaps by uploading it to a GitHub repo, or alternatively, by providing some example code snippets.  I truly appreciate any time and assistance you can offer. Thank you very much for considering my request.
Best wishes

**Comment by PTRRupprecht** [comment]
Hi @Achuan-2,
The code generated with GPT does indeed something that I can understand as dilated regression. It is already close to what I have written myself. The main difference is that your code only dilates toward future time points, whereas my version used dilation towards past and future time points.

Unfortunately, I don't have time right now to refactor the entire code. But here's a piece of code that might help you figure things out. The `indices1` and `indices2` average across the past and future values of the current dilation bin. `Temp_dilated_regressor` is the regressor for the current time bin (indexed here by the parameter `j`). If you don't manage to figure things out using your already good starting point and this little piece, let me know!

```window_size = 512
exponent = log2(window_size) - 1;
center_index = window_size/2 + j; % current time point of the regressor
clear tempp
for ggg = 1:exponent
    indices1 = (center_index + 2^(ggg-1) - 1):(center_index + 2^ggg - 2);
    indices2 = ((center_index - 2^ggg + 2):(center_index - 2^(ggg-1) + 1) ) - 1;
    Temp_dilated_regressor(ggg + exponent ) = nanmean(X_regressor(indices1));
    Temp_dilated_regressor(exponent + 1 - ggg ) = nanmean(X_regressor(indices2));
end
```

I also have to admit that I don't know the `regress()` function in Matlab too well. I guess it will work here. I have used the function `fitglm()`. In the end, the result will be probably very similar or identical.

Best wishes,
Peter

**Comment by Achuan-2** [comment]
Hi @PTRRupprecht 
I seem to know how to do the dilated regression. Based on my understanding, I have drawn a schematic diagram: 
![Clip_2024-04-22_01-22-11](https://github.com/HelmchenLabSoftware/Centripetal_propagation_astrocytes/assets/60436214/115279d5-412d-4aca-b573-46fe779138e7)

And the following is the code generated by GPT:
```matlab
% Assume X_regressor is our independent variable, and Y is the dependent variable
% Assume our data sampling rate is 30Hz, with an effective time window of ±17s
% Thus, window_size is 512 (the number of frames corresponding to ±17s)

% Initialize parameters
window_size = 512;
sample_rate = 30; % 30Hz
total_timepoints = length(X_regressor); % Total number of time points
exponent = log2(window_size) - 1; % Calculate exponent based on window size

% Initialize the matrix for dilated regressors
dilated_regressors = nan(total_timepoints, 2 * exponent + 1);

% Generate dilated regressors
for j = 1:total_timepoints
    center_index = window_size/2 + j; % Index of the current time point
    for ggg = 1:exponent
        indices1 = (center_index + 2^(ggg-1) - 1):(center_index + 2^ggg - 2);
        indices2 = ((center_index - 2^ggg + 2):(center_index - 2^(ggg-1) + 1)) - 1;
        % Ensure indices are within valid range
        indices1 = indices1(indices1 > 0 & indices1 <= total_timepoints);
        indices2 = indices2(indices2 > 0 & indices2 <= total_timepoints);
        % Calculate mean value for the dilated regressor
        dilated_regressors(j, ggg + exponent) = nanmean(X_regressor(indices1));
        dilated_regressors(j, exponent + 1 - ggg) = nanmean(X_regressor(indices2));
    end
end

% Now that we have the dilated regressors and the dependent variable Y, we can perform linear regression
% Use fitlm function for linear regression instead of regress
lm = fitlm(dilated_regressors, Y); % fitlm function provides a linear model

% Now, 'lm' contains the linear model, we can use it to predict new dependent variable values
Y_pred = predict(lm, dilated_regressors);

% Calculate the effectiveness of the prediction, for example, by calculating the coefficient of determination R^2
SStot = sum((Y - mean(Y)).^2);
SSres = sum((Y - Y_pred).^2);
R_squared = 1 - (SSres/SStot);

% Output results
fprintf('R-squared for dilated linear regression: %.2f\n', R_squared);

```
Could you tell me that my understanding is correct? Thank you very much for guidance!
Best Wishes!

**Comment by PTRRupprecht** [comment]
Hi @Achuan-2,

Yes, from going through the code, I think it should work - looks great!

The only thing that you may have to be changed for a real application is the fitting procedure itself. Currently, you cannot see whether overfitting does happen or not. To do so, you have to perform cross-validation (in our paper, I did 5-fold cross-validation). Basically, you hold back part of the data for validation, fit the regression on the rest and report only the performance on the validation set. But probably this is already clear to you.

Best,
Peter

**Comment by Achuan-2** [comment]
OK. It is a pleasure discussing with you.That's very kind of you. Thanks a lot.
May your research be fruitful and your life be filled with joy and happiness.
